CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,226 @@
+```cddl
+transaction_unspent_output = [
+  input: transaction_input,
+  output: transaction_output,
+]
+```
+then we define
+```
+type TransactionUnspentOutput = cbor<transaction_unspent_output>
+```
+
+This allows us to use the output for constructing new transactions using it as an output as the `transaction_output` in the Shelley Mary CDDL does not contain enough information on its own to spend it.
+
+### Paginate
+
+```
+type Paginate = {|
+  page: number,
+  limit: number,
+|};
+```
+Used to specify optional pagination for some API calls. Limits results to {limit} each page, and uses a 0-indexing {page} to refer to which of those pages of {limit} items each.
+
+
+## Error Types
+
+### APIError
+
+```
+APIErrorCode {
+	InvalidRequest: -1,
+	InternalError: -2,
+	Refused: -3,
+}
+APIError {
+	code: APIErrorCode,
+	info: string
+}
+```
+
+### DataSignError
+
+```
+DataSignErrorCode {
+	ProofGeneration: 1,
+	AddressNotPK: 2,
+	UserDeclined: 3,
+	InvalidFormat: 4,
+}
+type DataSignError = {
+	code: DataSignErrorCode,
+	info: String
+}
+```
+
+* ProofGeneration - Wallet could not sign the data (e.g. does not have the secret key associated with the address)
+* AddressNotPK - Address was not a P2PK address and thus had no SK associated with it.
+* UserDeclined - User declined to sign the data
+* InvalidFormat - If a wallet enforces data format requirements, this error signifies that the data did not conform to valid formats.
+
+* InvalidRequest - Inputs do not conform to this spec or are otherwise invalid.
+* InternalError - An error occurred during execution of this API call.
+* Refused - The request was refused due to lack of access - e.g. wallet disconnects.
+
+### PaginateError
+
+```
+type PaginateError = {|
+    maxSize: number,
+|};
+```
+{maxSize} is the maximum size for pagination and if the dApp tries to request pages outside of this boundary this error is thrown.
+
+### TxSendError
+
+```
+TxSendErrorCode = {
+	Refused: 1,
+	Failure: 2,
+}
+type TxSendError = {
+	code: TxSendErrorCode,
+	info: String
+}
+```
+
+* Refused - Wallet refuses to send the tx (could be rate limiting)
+* Failure - Wallet could not send the tx
+
+### TxSignError
+
+```
+TxSignErrorCode = {
+	ProofGeneration: 1,
+	UserDeclined: 2,
+}
+type TxSignError = {
+	code: TxSignErrorCode,
+	info: String
+}
+```
+
+* ProofGeneration - User has accepted the transaction sign, but the wallet was unable to sign the transaction (e.g. not having some of the private keys)
+* UserDeclined - User declined to sign the transaction
+
+
+
+## Initial API
+
+In order to initiate communication from webpages to a user's Cardano wallet, the wallet must provide the following javascript functions to the webpage. These would likely be done via code injection into the webpage. The API is split into two stages to maintain the user's privacy, as the user will have to consent to `cardano_request_read_access()` in order for the dApp to read any information pertaining to the user's wallet. 
+
```

Comment:
Exactly. The `cardano` object is first of all injected if the `cardano_requests_read_access` function returns true.

Reply:
We did this two-phase so that the cardano object was separate and if it exists then it meant that the rest of the API was injected. It was originally all in there e.g. `cardano.request_read_access()` but we changed the decision later. However, if the user disconnects the wallet then dApps still need to handle `cardano.*()` returning an error that the wallet is not connected, so it's possible we could revert to that behavior.

At page load those 2 functions are injected, and if `cardano_request_read_access()` is successful then the `cardano` object is injected with the rest of the API as alessandro mentioned.

Reply:
Another idea I had thought of previously was if `cardano_request_read_access()` returns a `Cardano` instance instead of injecting it as `global.cardano`. We would still need to inject the type definition if it's not injected at wallet start, and if it is at wallet start we would want to control in the implementation to not allow dApps creating their own object from that type and trying to use it.

Reply:
> It was originally all in there e.g. cardano.request_read_access() but we changed the decision later. 

I'd be curious to know why, to me it seems more consistent to keep everything under the same namespace. 

> Another idea I had thought of previously was if cardano_request_read_access() returns a Cardano instance 

I like this idea very much :+1:

Reply:
>I'd be curious to know why, to me it seems more consistent to keep everything under the same namespace.

Well at first I had wanted to simplify things so that dApps wouldn't have access to those APIs until they had been granted access without having to check for API access denied errors in every endpoint, but it's not really a valid point once you realize that you'll need to handle those anyway if a wallet cancels access later on after injection. It does let us potentially go with that returning the `cardano` object idea instead of performing a 2nd injection so that was a consideration as well.

Reply:
Yeah I think just hiding the endpoints, doesn't fully prevent someone from accessing the wallet actually, since you communicate with window.postMessage. so the prevention needs to be in the extension itself. I would also prefer to move it under the same name space. maybe make it shorter like:
`cardano.enable` and `cardano.is_enabled`

Reply:
@alessandrokonrad It absolutely doesn't on its own. That is up to the wallet's implementation. When we implemented the ergo dApp connector spec in Yoroi we handle the connections from inside the Yoroi extension so Yoroi will only only respond to requests from pages that it considers connected, which can only happen if the connect request was accepted by the user (either by whitelisting the site or by the popup). You can't assume much at all in the way of security guarantees whatsoever in the injected code. That's why we treat the `yoroi-ergo-connector` (especially the injected code) as a trustless relay, and inside of `yoroi-frontend` we verify the inputs and keep track of connections.

Reply:
Concerning this, we could potentially expand on this and other implementation-related security issues in the spec, but it might be outside of the scope.

Reply:
I changed how these things work while I was reworking the wallet connection UX for when you have multiple wallets (e.g. Yoroi + AdaLite) that need to inject their API.